src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs

```diff
@@ -1864,12 +1944,63 @@ private static BaseValueTypeValue NewUninitializedLocationValue(TypeDesc locatio
+        private bool TryHandleIntrinsicCall(TypeSystemContext context, MethodDesc method, Value[] parameters, out Value retVal)
         {
             retVal = default;
 
             switch (method.Name)
             {
+                case "Memmove":
+                    if (method.OwningType is MetadataType spanHelpersType
+                        && spanHelpersType.Name == "SpanHelpers" && spanHelpersType.Namespace == "System"
+                        && spanHelpersType.Module == spanHelpersType.Context.SystemModule
+                        && parameters[0] is ByRefValue dest
+                        && parameters[1] is ByRefValue src
+                        && parameters[2] is ValueTypeValue len)
+                    {
+                        int length;
+                        if (context.Target.PointerSize == 8)
+                        {
+                            long longLength = len.AsInt64();
+                            if (longLength > int.MaxValue || longLength < int.MinValue)
+                            {
+                                return false;
+                            }
+
+                            length = (int)longLength;
+                        }
+                        else
+                        {
+                            length = len.AsInt32();
+                        }
+                        var srcSpan = new Span<byte>(src.PointedToBytes, src.PointedToOffset, length);
+                        var dstSpan = new Span<byte>(dest.PointedToBytes, dest.PointedToOffset, length);
+                        srcSpan.CopyTo(dstSpan);
+                        return true;
+                    }
+                    return false;
+                case "IsReferenceOrContainsReferences":
+                    if (method.OwningType is MetadataType rtType
+                        && rtType.Name == "RuntimeHelpers" && rtType.Namespace == "System.Runtime.CompilerServices"
+                        && rtType.Module == rtType.Context.SystemModule
+                        && method.Instantiation.Length == 1)
+                    {
+                        var type = method.Instantiation[0];
+                        bool result = type.IsGCPointer || (type is DefType { ContainsGCPointers: true });
+                        retVal = ValueTypeValue.FromSByte(result ? (sbyte)1 : (sbyte)0);
+                        return true;
+                    }
+                    return false;
+                case "FastAllocateString":
+                    if (method.OwningType is MetadataType stringType
+                        && stringType.Name == "String" && stringType.Namespace == "System"
+                        && stringType.Module == stringType.Context.SystemModule
+                        && parameters[0] is ValueTypeValue strSize)
+                    {
+                        retVal = new StringInstance(context.GetWellKnownType(WellKnownType.String), new string('\0', strSize.AsInt32()));
```

Comment:
Do we need a limit on the string size that can be allocated here? We do not want to OOM the compiler if somebody happens to create extremely long strings in the static constructors.